File: derp/credits/service.py
"""Credit service for checking and deducting credits.

This is the main entry point for credit operations. It handles:
- Chat model selection based on credit balance
- Tool access checking with daily limits
- Credit deduction after successful operations
- Credit purchases
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import TYPE_CHECKING

import logfire

from derp.catalog import GoogleModelKey
from derp.credits.tools import TOOL_REGISTRY, get_tool
from derp.credits.types import CreditCheckResult
from derp.db.credits import (
    add_chat_credits,
    add_user_credits,
    deduct_chat_credits,
    deduct_user_credits,
    get_balances,
    get_daily_usage,
    get_transaction_by_idempotency_key,
    increment_daily_usage,
)
from derp.execution import ExecutionPlan, Feature, plan_execution
from derp.history.service import HISTORY_WINDOWS, HistoryWindow
from derp.observability import telemetry_fingerprint

if TYPE_CHECKING:
    from sqlalchemy.ext.asyncio import AsyncSession

    from derp.models import Chat, User


class CreditService:
    """Service for managing credits and access control.

    Methods accept database models (User, Chat) directly for cleaner API.

    Usage:
        service = CreditService(session)

        # Get the exact orchestrator plan and context policy.
        plan, history_window = await service.get_orchestrator_config(user, chat)

        # Check tool access
        result = await service.check_tool_access(user, chat, "image_generate")
        if result.allowed:
            # Execute tool
            ...
            # Deduct credits after success
            await service.deduct(result, user, chat, "image_generate")
    """

    def __init__(self, session: AsyncSession) -> None:
        self.session = session

    async def get_orchestrator_config(
        self,
        user: User,
        chat: Chat,
    ) -> tuple[ExecutionPlan, HistoryWindow]:
        """Get orchestrator configuration based on credit balance.

        Args:
            user: Database User model.
            chat: Database Chat model.

        Returns:
            Validated chat execution plan plus its product history window.
        """
        chat_credits, user_credits = await get_balances(
            self.session, user.telegram_id, chat.telegram_id
        )

        if chat_credits > 0 or user_credits > 0:
            model_key = GoogleModelKey.CHAT_STANDARD
        else:
            model_key = GoogleModelKey.CHAT_ECONOMY

        plan = plan_execution(Feature.CHAT, model_key)
        history_window = HISTORY_WINDOWS[plan.model.key]

        logfire.debug(
            "orchestrator_config",
            model_key=plan.model.key.value,
            model=plan.model.provider_model_id,
            history_max_turns=history_window.max_turns,
            history_max_tokens=history_window.max_tokens,
            chat_credits=chat_credits,
            user_credits=user_credits,
        )

        return plan, history_window

    async def check_tool_access(
        self,
        user: User,
        chat: Chat,
        tool_name: str,
        *,
        arguments: Mapping[str, object] | None = None,
    ) -> CreditCheckResult:
        """Check if a tool can be used, considering credits and daily limits.

        The check order:
        1. Free daily limit (if available)
        2. Chat credits
        3. User credits
        4. Reject

        Args:
            user: Database User model.
            chat: Database Chat model.
            tool_name: Name of the tool to check.
            arguments: Validated tool arguments used for model variants.

        Returns:
            CreditCheckResult with access decision and details.
        """
        tool = get_tool(tool_name)
        resolved_arguments = arguments or {}
        plan = tool.resolve_plan(resolved_arguments)
        total_cost = tool.total_cost(tool.model_credit_cost(plan, resolved_arguments))

        # Get balances
        chat_credits, user_credits = await get_balances(
            self.session, user.telegram_id, chat.telegram_id
        )

        # Check free daily limit first
        if tool.free_daily_limit > 0:
            used = await get_daily_usage(self.session, user.id, chat.id, tool_name)
            if used < tool.free_daily_limit:
                return CreditCheckResult(
                    allowed=True,
                    plan=plan,
                    source="free",
                    credits_to_deduct=0,
                    credits_remaining=None,
                    free_remaining=tool.free_daily_limit - used - 1,
                )

        if total_cost == 0:
            return CreditCheckResult(
                allowed=False,
                plan=plan,
                source="rejected",
                credits_to_deduct=0,
                credits_remaining=None,
                free_remaining=0,
                reject_reason=f"Daily limit reached for {tool.name}",
            )

        # Check chat credits
        if chat_credits >= total_cost:
            return CreditCheckResult(
                allowed=True,
                plan=plan,
                source="chat",
                credits_to_deduct=total_cost,
                credits_remaining=chat_credits - total_cost,
                free_remaining=None,
            )

        # Check user credits
        if user_credits >= total_cost:
            return CreditCheckResult(
                allowed=True,
                plan=plan,
                source="user",
                credits_to_deduct=total_cost,
                credits_remaining=user_credits - total_cost,
                free_remaining=None,
            )

        # Rejected
        return CreditCheckResult(
            allowed=False,
            plan=plan,
            source="rejected",
            credits_to_deduct=0,
            credits_remaining=0,
            free_remaining=0,
            reject_reason=f"Need {total_cost} credits for {tool.name}",
        )

    async def deduct(
        self,
        result: CreditCheckResult,
        user: User,
        chat: Chat,
        tool_name: str,
        *,
        idempotency_key: str | None = None,
        metadata: dict | None = None,
    ) -> None:
        """Deduct credits after successful tool execution.

        Call this ONLY after the tool has successfully completed.
        Uses idempotency_key to prevent double-charging on retries.

        Args:
            result: The CreditCheckResult from check_tool_access.
            user: Database User model.
            chat: Database Chat model.
            tool_name: Name of the tool used.
            idempotency_key: Optional key to prevent duplicate charges.
            metadata: Optional additional context.
        """
        if not result.allowed:
            raise ValueError("Cannot deduct a rejected credit check")
        tool = get_tool(tool_name)
        if (result.plan is None) != (tool.feature is None):
            raise ValueError(f"{tool_name} received an incompatible execution plan")
        if result.plan and result.plan.feature is not tool.feature:
            raise ValueError(
                f"{result.plan.feature.value} cannot settle as {tool_name}"
            )

        # Check idempotency
        if idempotency_key:
            existing = await get_transaction_by_idempotency_key(
                self.session, idempotency_key
            )
            if existing:
                logfire.info(
                    "deduction_skipped_idempotent",
                    idempotency_key=idempotency_key,
                    tool=tool_name,
                )
                return

        if result.source == "free":
            await increment_daily_usage(self.session, user.id, chat.id, tool_name)
            logfire.info(
                "free_usage_incremented",
                tool=tool_name,
                user_id=str(user.id),
                chat_id=str(chat.id),
            )
        elif result.source == "chat":
            await deduct_chat_credits(
                self.session,
                chat.id,
                user.id,
                result.credits_to_deduct,
                tool_name,
                result.model_id,
                idempotency_key=idempotency_key,
                metadata=metadata,
            )
            logfire.info(
                "chat_credits_deducted",
                amount=result.credits_to_deduct,
                tool=tool_name,
                chat_id=str(chat.id),
            )
        elif result.source == "user":
            await deduct_user_credits(
                self.session,
                user.id,
                result.credits_to_deduct,
                tool_name,
                result.model_id,
                idempotency_key=idempotency_key,
                metadata=metadata,
            )
            logfire.info(
                "user_credits_deducted",
                amount=result.credits_to_deduct,
                tool=tool_name,
                user_id=str(user.id),
            )

    async def purchase_credits(
        self,
        user: User,
        chat: Chat | None,
        amount: int,
        telegram_charge_id: str,
        *,
        pack_name: str | None = None,
    ) -> int:
        """Process a credit purchase from Telegram Stars.

        Args:
            user: Database User model.
            chat: Database Chat model (None for personal credits).
            amount: Number of credits to add.
            telegram_charge_id: Telegram payment charge ID (for idempotency).
            pack_name: Optional pack name for metadata.

        Returns:
            New balance after purchase.
        """
        # Use telegram_charge_id as idempotency key
        existing = await get_transaction_by_idempotency_key(
            self.session, telegram_charge_id
        )
        if existing:
            logfire.info(
                "purchase_skipped_duplicate",
                charge_fingerprint=telemetry_fingerprint(telegram_charge_id),
            )
            # Return the balance from the existing transaction
            return existing.balance_after

        metadata = {"pack_name": pack_name} if pack_name else {}

        if chat:
            new_balance = await add_chat_credits(
                self.session,
                chat.id,
                user.id,
                amount,
                "purchase",
                telegram_charge_id=telegram_charge_id,
                idempotency_key=telegram_charge_id,
                metadata=metadata,
            )
            logfire.info(
                "chat_credits_purchased",
                amount=amount,
                chat_id=str(chat.id),
                user_id=str(user.id),
                new_balance=new_balance,
            )
        else:
            new_balance = await add_user_credits(
                self.session,
                user.id,
                amount,
                "purchase",
                telegram_charge_id=telegram_charge_id,
                idempotency_key=telegram_charge_id,
                metadata=metadata,
            )
            logfire.info(
                "user_credits_purchased",
                amount=amount,
                user_id=str(user.id),
                new_balance=new_balance,
            )

        return new_balance

    async def refund_credits(
        self,
        telegram_charge_id: str,
    ) -> bool:
        """Process a refund for a previous purchase.

        Finds the original transaction and reverses it.

        Args:
            telegram_charge_id: The original purchase's charge ID.

        Returns:
            True if refund was processed, False if original not found.
        """
        original = await get_transaction_by_idempotency_key(
            self.session, telegram_charge_id
        )
        if not original:
            logfire.warn(
                "refund_failed_not_found",
                charge_fingerprint=telemetry_fingerprint(telegram_charge_id),
            )
            return False

        if original.type != "purchase":
            logfire.warn(
                "refund_failed_not_purchase",
                charge_fingerprint=telemetry_fingerprint(telegram_charge_id),
                type=original.type,
            )
            return False

        # Reverse the transaction
        refund_key = f"refund:{telegram_charge_id}"
        existing_refund = await get_transaction_by_idempotency_key(
            self.session, refund_key
        )
        if existing_refund:
            logfire.info(
                "refund_already_processed",
                charge_fingerprint=telemetry_fingerprint(telegram_charge_id),
            )
            return True

        if original.chat_id:
            await add_chat_credits(
                self.session,
                original.chat_id,
                original.user_id,
                -original.amount,  # Negative to remove
                "refund",
                idempotency_key=refund_key,
                metadata={"original_charge_id": telegram_charge_id},
            )
        else:
            await add_user_credits(
                self.session,
                original.user_id,
                -original.amount,
                "refund",
                idempotency_key=refund_key,
                metadata={"original_charge_id": telegram_charge_id},
            )

        logfire.info(
            "refund_processed",
            charge_fingerprint=telemetry_fingerprint(telegram_charge_id),
            amount=original.amount,
        )
        return True


def get_placeholder_message(tool_name: str, reject_reason: str) -> str:
    """Generate a placeholder message for tools that can't be used.

    This message is returned to the agent so it can naturally respond
    to the user about the limitation.
    """
    tool = TOOL_REGISTRY.get(tool_name)
    if not tool:
        return f"[TOOL_UNAVAILABLE: {reject_reason}]"

    return (
        f"[TOOL_UNAVAILABLE: {tool.description} requires credits. "
        f"{reject_reason}. Do not suggest a purchase; credit purchases are unavailable.]"
    )

File: src/credits/service.ts
import { config } from "../config";
import type { Database } from "../db/connection";
import {
	addChatCredits,
	addUserCredits,
	deductChatCredits,
	deductUserCredits,
	getBalances,
	getTransactionByIdempotencyKey,
	type ToolDebitResult,
} from "../db/queries/credits";
import {
	getOpenDebtAmount,
	getQuotaUsage,
	recordGlobalFreeToolUsage,
} from "../db/queries/finance";
import type { Chat, User } from "../db/schema";
import {
	CONTEXT_LIMITS,
	getDefaultModel,
	getModel,
	ModelCapability,
	ModelTier,
} from "../llm/registry";
import type { CreditCheckResult } from "./types";

// ── Tool pricing lookup (kept in sync with ToolRegistry) ─────────────────────

interface ToolPricing {
	credits: number;
	freeDaily: number;
	capability?: ModelCapability;
	defaultModel?: string;
}

const toolPricingMap = new Map<string, ToolPricing>();
export const STANDARD_CHAT_TOOL_NAME = "chat";
export const STANDARD_CHAT_CREDITS = 2;

toolPricingMap.set(STANDARD_CHAT_TOOL_NAME, {
	credits: STANDARD_CHAT_CREDITS,
	freeDaily: 0,
	capability: ModelCapability.TEXT,
});

/** Register a tool's pricing (called by ToolRegistry on startup) */
export function registerToolPricing(name: string, pricing: ToolPricing): void {
	toolPricingMap.set(name, pricing);
}

export function hasActiveSubscription(
	user: Pick<User, "subscriptionTier" | "subscriptionExpiresAt">,
): boolean {
	return (
		user.subscriptionTier != null &&
		user.subscriptionExpiresAt != null &&
		user.subscriptionExpiresAt > new Date()
	);
}

function quotaScopeForTool(toolName: string): "web_search" | "promo_media" {
	return toolName === "webSearch" ? "web_search" : "promo_media";
}

function botWideLimitForTool(toolName: string): number {
	return toolName === "webSearch"
		? config.freeSearchDailyBotLimit
		: config.freePromoMediaDailyBotLimit;
}

// ── Credit Service ───────────────────────────────────────────────────────────

export class CreditService {
	constructor(
		private db: Database,
		private user: User,
		private chat: Chat,
	) {}

	/** Get orchestrator config: which model/tier to use for chat */
	async getOrchestratorConfig(): Promise<{
		tier: ModelTier;
		modelId: string;
		contextLimit: number;
	}> {
		const { userCredits, chatCredits } = await getBalances(
			this.db,
			this.user.telegramId,
			this.chat.telegramId,
		);
		const [userDebt, chatDebt] = await Promise.all([
			getOpenDebtAmount(this.db, { userId: this.user.id }),
			getOpenDebtAmount(this.db, {
				userId: this.user.id,
				chatId: this.chat.id,
			}),
		]);

		const hasPaid =
			(chatCredits >= STANDARD_CHAT_CREDITS && chatDebt === 0) ||
			(userCredits >= STANDARD_CHAT_CREDITS && userDebt === 0);
		const tier = hasPaid ? ModelTier.STANDARD : ModelTier.FREE;
		const model = getDefaultModel(ModelCapability.TEXT, tier);
		const contextLimit = CONTEXT_LIMITS[tier];

		return { tier, modelId: model.id, contextLimit };
	}

	/** Check if a tool can be used */
	async checkToolAccess(toolName: string): Promise<CreditCheckResult> {
		const pricing = toolPricingMap.get(toolName);
		if (!pricing) {
			return {
				allowed: false,
				tier: ModelTier.FREE,
				modelId: "",
				source: "rejected",
				creditsToDeduct: 0,
				creditsRemaining: 0,
				freeRemaining: 0,
				rejectReason: `Unknown tool: ${toolName}`,
			};
		}

		// Resolve model
		let model: ReturnType<typeof getModel>;
		if (pricing.defaultModel) {
			model = getModel(pricing.defaultModel);
		} else if (pricing.capability) {
			model = getDefaultModel(pricing.capability, ModelTier.STANDARD);
		} else {
			model = getDefaultModel(ModelCapability.TEXT, ModelTier.STANDARD);
		}

		const totalCost = pricing.credits;
		const hasMeteredFreeQuota =
			Number.isFinite(pricing.freeDaily) && pricing.freeDaily > 0;

		// Check free daily limit first
		if (hasMeteredFreeQuota) {
			const scope = quotaScopeForTool(toolName);
			const [used, botWideUsed] = await Promise.all([
				getQuotaUsage(this.db, {
					scope,
					userId: this.user.id,
				}),
				getQuotaUsage(this.db, {
					scope,
					botWide: true,
				}),
			]);
			const botWideLimit = botWideLimitForTool(toolName);
			if (used < pricing.freeDaily && botWideUsed < botWideLimit) {
				return {
					allowed: true,
					tier: model.tier,
					modelId: model.id,
					source: "free",
					creditsToDeduct: 0,
					creditsRemaining: null,
					freeRemaining: pricing.freeDaily - used - 1,
				};
			}
		}

		if (totalCost === 0 && hasMeteredFreeQuota) {
			return {
				allowed: false,
				tier: model.tier,
				modelId: model.id,
				source: "rejected",
				creditsToDeduct: 0,
				creditsRemaining: 0,
				freeRemaining: 0,
				rejectReason: `Daily free limit reached for ${toolName}`,
			};
		}

		// If tool is completely free and unmetered, always allow
		if (totalCost === 0) {
			return {
				allowed: true,
				tier: model.tier,
				modelId: model.id,
				source: "free",
				creditsToDeduct: 0,
				creditsRemaining: null,
				freeRemaining: null,
			};
		}

		// Check balances
		const { userCredits, chatCredits } = await getBalances(
			this.db,
			this.user.telegramId,
			this.chat.telegramId,
		);
		const [userDebt, chatDebt] = await Promise.all([
			getOpenDebtAmount(this.db, { userId: this.user.id }),
			getOpenDebtAmount(this.db, {
				userId: this.user.id,
				chatId: this.chat.id,
			}),
		]);

		// Chat credits first
		if (chatCredits >= totalCost) {
			if (chatDebt === 0) {
				return {
					allowed: true,
					tier: model.tier,
					modelId: model.id,
					source: "chat",
					creditsToDeduct: totalCost,
					creditsRemaining: chatCredits - totalCost,
					freeRemaining: null,
				};
			}
		}

		// User credits
		if (userCredits >= totalCost) {
			if (userDebt > 0) {
				return {
					allowed: false,
					tier: model.tier,
					modelId: model.id,
					source: "rejected",
					creditsToDeduct: 0,
					creditsRemaining: userCredits,
					freeRemaining: null,
					rejectReason: `You have ${userDebt} refund-debt credits to settle before paid tools can use personal credits`,
				};
			}
			return {
				allowed: true,
				tier: model.tier,
				modelId: model.id,
				source: "user",
				creditsToDeduct: totalCost,
				creditsRemaining: userCredits - totalCost,
				freeRemaining: null,
			};
		}

		return {
			allowed: false,
			tier: model.tier,
			modelId: model.id,
			source: "rejected",
			creditsToDeduct: 0,
			creditsRemaining: 0,
			freeRemaining: 0,
			rejectReason: `Need ${totalCost} credits for ${toolName}`,
		};
	}

	/** Check whether an idempotent tool request already has a ledger entry. */
	async hasProcessed(idempotencyKey: string): Promise<boolean> {
		return (
			(await getTransactionByIdempotencyKey(this.db, idempotencyKey)) != null
		);
	}

	/** Reserve or record tool credits before provider work. */
	async deduct(
		result: CreditCheckResult,
		toolName: string,
		idempotencyKey?: string,
		meta?: Record<string, unknown>,
	): Promise<ToolDebitResult> {
		// Check idempotency
		if (idempotencyKey) {
			const existing = await getTransactionByIdempotencyKey(
				this.db,
				idempotencyKey,
			);
			if (existing) return "duplicate";
		}

		if (result.source === "free") {
			const pricing = toolPricingMap.get(toolName);
			if (
				pricing &&
				Number.isFinite(pricing.freeDaily) &&
				pricing.freeDaily > 0
			) {
				return await recordGlobalFreeToolUsage(this.db, {
					userId: this.user.id,
					chatId: this.chat.id,
					toolName,
					modelId: result.modelId,
					limit: pricing.freeDaily,
					botWideLimit: botWideLimitForTool(toolName),
					idempotencyKey,
					meta,
				});
			}
		} else if (result.source === "chat") {
			const debit = await deductChatCredits(
				this.db,
				this.chat.id,
				this.user.id,
				result.creditsToDeduct,
				toolName,
				result.modelId,
				idempotencyKey,
				meta,
			);
			return debit.applied ? "applied" : "duplicate";
		} else if (result.source === "user") {
			const debit = await deductUserCredits(
				this.db,
				this.user.id,
				result.creditsToDeduct,
				toolName,
				result.modelId,
				idempotencyKey,
				meta,
			);
			return debit.applied ? "applied" : "duplicate";
		}
		return "applied";
	}

	/** Refund a paid pre-execution reservation after a tool reports failure. */
	async refundDeduction(
		result: CreditCheckResult,
		toolName: string,
		idempotencyKey?: string,
		meta?: Record<string, unknown>,
	): Promise<void> {
		if (result.creditsToDeduct <= 0) return;
		const refundKey = idempotencyKey ? `${idempotencyKey}:refund` : undefined;
		const refundMeta = {
			...meta,
			reason: "tool_failed_after_reservation",
			toolName,
			modelId: result.modelId,
		};

		if (result.source === "chat") {
			await addChatCredits(
				this.db,
				this.chat.id,
				this.user.id,
				result.creditsToDeduct,
				"refund",
				undefined,
				refundKey,
				refundMeta,
			);
			return;
		}

		if (result.source === "user") {
			await addUserCredits(
				this.db,
				this.user.id,
				result.creditsToDeduct,
				"refund",
				undefined,
				refundKey,
				refundMeta,
			);
		}
	}
}
